### ISSUE TITLE
Perf questions with implementation

### ISSUE THREAD

**Issue by isiahmeadows** [planttheidea/fast-copy#7]
I noticed you were trying to write a super fast object-related library, so I decided to poke around a little bit in the implementation. [I created my own such library, too](https://github.com/isiahmeadows/clean-match/blob/master/clean-match.js) but for deep comparison instead. (BTW, I don't have benchmarks, but that's because I just forgot to save the Lodash/etc. benchmarks before actually publishing my initial commit. It's about 30-50% as fast as Lodash while doing about 4x as much.)

I did come across a few things that struck me as odd, so I have a few questions:

1. [Is there a reason why you're using the boolean accessors directly instead of `RegExp.prototype.flags` here?](https://github.com/planttheidea/fast-copy/blob/master/src/utils.js#L27-L52) I would normally use the boolean accessors, too, except if I'm immediately plugging it into `new RegExp(re.source, re.flags)`. If you just inlined it as `regExp.flags` [here](https://github.com/planttheidea/fast-copy/blob/master/src/utils.js#L226), you *should* see some speed boost here.
1. In [this particular case](https://github.com/planttheidea/fast-copy/blob/master/src/utils.js#L63-L66), is there a reason why you didn't fork it out into two separate functions? (Engines know how to inline that.)
    - Conversely, why not inline this in [the only caller](https://github.com/planttheidea/fast-copy/blob/master/src/utils.js#L200-L210), guarded with `HAS_PROPERTY_SYMBOL_SUPPORT`? You can avoid some unnecessary work here by just iterating the original array as you go. It's a cheap test and the common case is that either *all* are enumerable or *none* are, so I'm not sure if anything here could be gained even by filtering in place first. And with the guard, engines will typically strip out the dead code when it goes to optimize it, and even if it doesn't, CPUs will predict it correctly.
1. In [your map/set copying routine](https://github.com/planttheidea/fast-copy/blob/master/src/utils.js#L148-L172), have you considered just splitting the function into two? Engines will occasionally do loop-invariant code motion, but they don't typically in my experience do that *through* function barriers. Also, you would make it easier on the engine to discern how they're used, without hardly increasing code size much.
1. In your [typed array copying routine](https://github.com/planttheidea/fast-copy/blob/master/src/utils.js#L233-L242), that's not actually copying the buffer. That's just copying the view itself. You need to copy the buffer, too, if you want to do a true deep copy.
1. [Why aren't you lifting this out and adding the `realm` and `cache` as new parameters?](https://github.com/planttheidea/fast-copy/blob/master/src/index.js#L28-L86)

If this helps for other things, here's a dump of some what I found with performance when dealing with highly polymorphic and allocaton-heavy stuff:

1. I would say the best tool out there I've found for diagnosing low-level performance issues in Node is V8's Turbolizer, located in the `tools/turbolizer/` subdirectory of its source code. The instructions do generally work if you substitute `node` for `d8`, and it's a lifesaver when you want to know what the heck TurboFan is doing to your code.
1. If you can convert a polymorphic call to a monomorphic one without adding much code independent of this, try it, even if it means duplicating a little bit of code. Every engine out there with tracing ICs loves monomorphic calls - if it's not inlined, the call is still specialized for the types it's being passed, and you're likely to see perf wins this way.
1. If it's megamorphic and recursive, it doesn't usually get inlined, and in my experience, it hits the slow path of almost *everything*. Also, sometimes in cases like this, when the engine attempts to inline, it inlines the wrong thing and completely breaks your performance altogether. I've been bit with this more than once with clean-match, linked above.
1. When something is very naturally allocation-heavy like this, it's easy for extraneous allocations to look like noise - I had to deal with that in [Mithril](https://mithril.js.org/) quite a bit already (I'm a core committer there). It's in cases like these where the first major perf wins are just shedding the extra weight of objects that don't need to exist
1. Megamorphic calls are typically something to avoid, but you only run into most performance cliffs when you attempt to do something that requires checking an object's type map at runtime (like reading a property or writing a property). From doing my own experimentation, here's some of the things I found:
    - Identity checking (like `a === b`) and nullish checking (like `x == null`) are *fully* polymorphic, so it doesn't matter how polymorphic your types are. No types need checked here.
    - Things like `typeof x === "object"` perform a memory access and partial type map check (for heap objects) worst case scenario, which can thwart a few attempts to optimize, but they do *not* require a full property map check.
    - Prototype accesses via `Object.getPrototypeOf(x)` only require a partial type map check for heap objects, since they don't need to care about the underlying property map.
    - Property accesses and anything else of the sort *will* incur a full type map check, which involves sometimes a pretty expensive series of memory accesses from disparate memory locations that are almost never in cache.
    - If you guard a code path with `x != null`, an engine can assume property and prototype access is always possible, which enables a few optimizations.
    - If you guard a code path with `x !== null && typeof x === "object"`, an engine can assume you not only have a heap object, but a proper reference type. This simplifies quite a few things, like `Object.getPrototypeOf(x)`, which can likely be inlined in its most specific form. It also allows it to avoid also having to check the type of the heap object itself to know how to perform the property access. (Representations for things like doubles and strings are different than for normal JS objects, making this distinction pretty meaningful.)

**Comment by planttheidea** [comment]
Thank you for your interest in this library! I took a lot of your advice in the referenced PR, and I'll try to respond with as much thought as you put into your original post. :)

**RegExp.prototype.flags**

Excellent recommendation, I was unaware of this property. I still kept the existing function for legacy browser support (namely IE), but I also augmented that method with the `unicode` and `sticky` values.

**getSymbols**

Not entirely sure what you meant by forking it into two functions, but your point about simply inlining it was dead-on. I originally abstracted it for reuse and simple testing, but it ended up only being used in the context of `copyObject`, so I was able to easily modify the tests for the new inline code. This boosted the speed of copying objects noticeably (I imagine in avoiding the double-iteration you called out with the `filter` usage).

**Map / Set copy**

This was a bit more tricky, because I value clean code just as much as I value performant code, so a straight copy-pasta made me cringe a bit. I was able to modify the `copyIterable` method into an HOF that accepted a context-specific handler for either maps or sets, and therefore create more targeted `copyMap` and `copySet` methods. This should provide a small performance boost, while also avoiding the aforementioned copy-pasta.

**TypedArray copy**

Good catch, I just modified the method to leverage `copyArrayBuffer` for the buffer before passing it to the constructor.

**Avoiding the closure**

This is ... more tricky, and its because the API gets more convoluted from an implementation perspective. For example, right now you pass an object (and optional realm) and get a deep copy of the object back. Simple, clean, understandable. Let's say I make the `cache` an optional third parameter (to avoid calling within the closure as you mention) ... now I've avoided the closure, but also potentially created a new option (a custom cache). We all know that if you give devleopers a tool, someone somewhere will use it, so ... what if I purposefully want to use a custom cache? Do I now need to explicitly pass the realm to ensure the order of parameters remains constant? Do I pass undefined? Or, does the second parameter now become an object of options to account for this (and maybe more) extensible points in the future?

Basically, this has the downstream effect of making things much more complex from a code perspective, purely for the sake of speed. If the performance gains were significant, I would consider this, but it is not something I would take lightly and would want to take a good amount of time to figure out the *right* way to do it from a consumer's perspective. Inevitably, for all my libraries, I value DX above all else, so I would want to take the time to maintain that.

**Overall**

Your recommendations were very helpful, and implementing them improved the performance across the board, and especially improved the copying of plain objects (which is one of the most likely use-cases). Much appreciated!

**Comment by isiahmeadows** [comment]
> This was a bit more tricky, because I value clean code just as much as I value performant code, so a straight copy-pasta made me cringe a bit. I was able to modify the copyIterable method into an HOF that accepted a context-specific handler for either maps or sets, and therefore create more targeted copyMap and copySet methods. This should provide a small performance boost, while also avoiding the aforementioned copy-pasta.

My general rule of thumb is this: I default to duplication first, then abstracting after about the third or fourth time. (It's IMHO easier to maintain, too - it's less chasing you have to do.) I personally am not persuaded by the elegance argument, since it's literally one more line of non-whitespace code (see my code below), but to each their own.

As for your higher-order function thingy, I doubt it actually changed much - bench it before and after. What you converted it into was arguably worse than what you had before - instead of having polymorphic object types, you now have a polymorphic function call into a closure. (I have no end of criticisms for engines doing a poor job optimizing closure variable access, but I've yet to see much change on that front in the last few years.)

Just to clarify, I was talking something like this for [this segment](https://github.com/planttheidea/fast-copy/blob/master/src/utils.js#L151-L162):

```js
export const copyMap = (map, copy, realm) => {
  const newMap = new map.constructor();

  map.forEach((value, key) => newMap.set(key, copy(value, realm)));

  return newMap;
}

export const copySet = (set, copy, realm) => {
  const newSet = new set.constructor();

  set.forEach((value, key) => newSet.add(copy(value, realm)));

  return newSet;
}
```

**Comment by isiahmeadows** [comment]
One last nit, then I'll address most of the other points.

- [What's with the `Object.create` here?](https://github.com/planttheidea/fast-copy/blob/master/src/utils.js#L17) You could safely just strip that out, since it's pretty useless.

> Excellent recommendation, I was unaware of this property. I still kept the existing function for legacy browser support (namely IE), but I also augmented that method with the unicode and sticky values.

Oops...I forgot that it was an ES<strong><em>6</em></strong> feature. 😄

> Not entirely sure what you meant by forking it into two functions, but your point about simply inlining it was dead-on. I originally abstracted it for reuse and simple testing, but it ended up only being used in the context of copyObject, so I was able to easily modify the tests for the new inline code. This boosted the speed of copying objects noticeably (I imagine in avoiding the double-iteration you called out with the filter usage).

The double iteration wouldn't have contributed much - perf changes from splitting loops mostly depend on *what* you do in them and more so, *when* you break out of them. The perf gains you got were from no longer creating and throwing away the extra array, and it wasn't until recently that engines started optimizing native Array methods to even compete with Lodash.

> This is ... more tricky, and its because the API gets more convoluted from an implementation perspective. For example, right now you pass an object (and optional realm) and get a deep copy of the object back. [...]

Not something I suggested changing API-wise. For this bit, just thought I'd clarify what I was referring to when I was talking about removing the closure. It'd look something like this, pretty clear cut and well-defined:

```js
export default function copy(object, realm = global) {
  return handleCopy(object, realm, getNewCache());
}

// This goes in a new file imported by both here and `./utils.js` so you don't
// have to pass it around and so the circular dependency is correctly handled in
// Babel + Node/Browserify.
export function handleCopy(object, realm, cache) {
  if (!isObjectCopyable(object, cache)) {
    return object;
  }

  if (Array.isArray(object)) {
    cache.add(object);

    return copyArray(object, realm);
  }

  if (object.constructor === realm.Object) {
    cache.add(object);

    return copyObject(object, realm, true);
  }

  if (object instanceof realm.Date) {
    return new Date(object.getTime());
  }

  if (object instanceof realm.RegExp) {
    return copyRegExp(object, realm);
  }

  if (realm.Map && object instanceof realm.Map) {
    cache.add(object);

    return copyMap(object, realm);
  }

  if (realm.Set && object instanceof realm.Set) {
    cache.add(object);

    return copySet(object, realm);
  }

  if (realm.Buffer && realm.Buffer.isBuffer(object)) {
    return copyBuffer(object, realm);
  }

  if (realm.ArrayBuffer) {
    if (realm.ArrayBuffer.isView(object)) {
      return copyTypedArray(object);
    }

    if (object instanceof realm.ArrayBuffer) {
      return copyArrayBuffer(object);
    }
  }

  if (shouldObjectBeCopied(object, realm)) {
    cache.add(object);

    return copyObject(object, realm);
  }

  return object;
}
```

The performance implications of this would be very noticeable when you copy small and simple objects. Currently, you allocate a minimum of four things for every basic object clone: a weak map, a function closure, the clone itself, and the symbol array. Furthermore, every time you reference `realm` and `cache` within `handleCopy`, it turns into a heap memory access just to get the object. My fix would drop it to 3 by removing the function closure allocation, and remove the indirection when accessing the realm and cache variables. And for smaller objects, this is *very* substantial.

> [...] Simple, clean, understandable. Let's say I make the cache an optional third parameter (to avoid calling within the closure as you mention) ... now I've avoided the closure, but also potentially created a new option (a custom cache). We all know that if you give devleopers a tool, someone somewhere will use it, so ... what if I purposefully want to use a custom cache? Do I now need to explicitly pass the realm to ensure the order of parameters remains constant? Do I pass undefined? Or, does the second parameter now become an object of options to account for this (and maybe more) extensible points in the future?
>
> Basically, this has the downstream effect of making things much more complex from an implementor's perspective, for the sake of performance gains (how much is unknown). If the performance gains were significant, I would consider this, but it is not something I would take lightly and would want to take a good amount of time to figure out the right way to do it from a consumer's perspective. Inevitably, for all my libraries, I value DX above all else, so I would want to take the time to maintain that.

I wouldn't worry about cache extensibility until someone actually requests it, and nothing here really affects *how* you expose that option - you could really just use anything with the right adapter. If you're concerned about prematurely exposing it, you can see my code snippet above, or you can take a look at [what I did here](https://github.com/isiahmeadows/clean-match/blob/master/clean-match.js#L151-L157). And if developers are dumb enough to rely on undocumented internals like that in a dependency, that's their problem, not yours. 😉

And in this particular case, I don't think there's any real use for exposing a custom cache. It's not like you aren't already doing the right thing now, and you're not relying on external APIs, I/O, or anything else of the link.

**Comment by planttheidea** [comment]
So ... the tone of this exchange took a markedly different tone with these responses. I'm really not looking to have ideological discussions about what our specific rule for abstraction vs inlining is, nor whether it's our responsibility as developers to consider how other developers consuming our product could make use of it, nor other stylistic opinions about how we fundamentally approach code. Certainly not over github issues for a specific library.

I do very much appreciate your interest in this library, and if you wish to contribute to its improvement, I always welcome PRs.
